### PR TITLE
Horizontal filtering

### LINKED ISSUE #51: Add smoothing operator to the static B
Implement as a linear variable change and use bump for C^1/2 and C^1/2^T

### PR DESCRIPTION
**Pull request by guillaumevernieres**
closes #51 
- Addition of a linear var transform for horizontal filtering using bump.
- Usage example in a new dirac test, uses the pre-computed correlation operator (output has the wrong amplitude) 
- Caveat: filtering is not yet a bump option (minor issue!)

### CODE REVIEW

File: src/soca/Transforms/HorizConv/HorizConv.cc [hunk 1]
```diff
@@ -0,0 +1,53 @@
+/*
+ * (C) Copyright 2017-2018  UCAR.
```
Comment (aerorahul) [inline src/soca/Transforms/HorizConv/HorizConv.cc:2]: The year is 2019.

File: src/soca/Transforms/HorizConv/HorizConv.cc [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    cov2d_.multiply(dxm, dxa);
+  }
+  // -----------------------------------------------------------------------------
+  void HorizConv::multiplyInverseAD(const Increment & dxa, Increment & dxm)
+    const {
+    dxm = dxa;
+  }
+  // -----------------------------------------------------------------------------
+  void HorizConv::print(std::ostream & os) const {
+    os << "SOCA change variable";
```
Comment (aerorahul) [inline src/soca/Transforms/HorizConv/HorizConv.cc:50]: SOCA Horizontal Covariance

File: src/soca/Transforms/HorizConv/HorizConv.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
+/*
+ * (C) Copyright 2017-2018  UCAR.
```
Comment (aerorahul) [inline src/soca/Transforms/HorizConv/HorizConv.h:2]: 2019

File: src/soca/Transforms/HorizConv/HorizConv.h [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+#include <ostream>
+#include <string>
+#include <boost/scoped_ptr.hpp>
+#include <boost/shared_ptr.hpp>
+
+#include "oops/util/DateTime.h"
+#include "oops/util/Printable.h"
+#include "oops/base/Variables.h"
+#include "soca/Covariance/ErrorCovariance.h"
```
Comment (aerorahul) [inline src/soca/Transforms/HorizConv/HorizConv.h:19]: norm fails here.

File: src/soca/Covariance/soca_covariance_mod.F90 [hunk 5]
```diff
@@ -352,7 +352,7 @@ subroutine soca_bump_correlation(self, horiz_convol, geom, c_conf, domain)
     ! Compute convolution weight
     call horiz_convol%setup_online(nc0a,nl0,nv,nts,lon,lat,area,vunit,lmask)
 
-    if (horiz_convol%nam%new_nicas) then
+    !if (horiz_convol%nam%new_nicas) then
```
Comment (aerorahul) [inline src/soca/Covariance/soca_covariance_mod.F90:355]: put a TODO: remove in the future or something like that.

File: src/soca/Transforms/HorizConv/HorizConv.cc [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+#include "soca/Transforms/HorizConv/HorizConv.h"
+
+#include <ostream>
+#include <string>
+
+#include "oops/util/Logger.h"
+#include "eckit/config/Configuration.h"
+#include "soca/Increment/Increment.h"
+#include "soca/State/State.h"
+#include "soca/Geometry/Geometry.h"
```
Comment (aerorahul) [inline src/soca/Transforms/HorizConv/HorizConv.cc:17]: norm fails here.

File: src/soca/Transforms/HorizConv/HorizConv.h [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  class Configuration;
+}
+
+namespace soca {
+  class State;
+  class Increment;
+  class Geometry;
+
+// -----------------------------------------------------------------------------
+/// SOCA linear change of variable
```
Comment (aerorahul) [inline src/soca/Transforms/HorizConv/HorizConv.h:33]: copy/paste. SOCA horizontal convolution.

File: src/soca/Transforms/HorizConv/HorizConv.h [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+#ifndef SOCA_TRANSFORMS_HORIZCONV_HORIZCONV_H_
+#define SOCA_TRANSFORMS_HORIZCONV_HORIZCONV_H_
+
+#include <ostream>
+#include <string>
+#include <boost/scoped_ptr.hpp>
+#include <boost/shared_ptr.hpp>
+
+#include "oops/util/DateTime.h"
```
Comment (aerorahul) [inline src/soca/Transforms/HorizConv/HorizConv.h:16]: unused.

File: src/soca/Transforms/HorizConv/HorizConv.h [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+ * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+#ifndef SOCA_TRANSFORMS_HORIZCONV_HORIZCONV_H_
+#define SOCA_TRANSFORMS_HORIZCONV_HORIZCONV_H_
+
+#include <ostream>
+#include <string>
+#include <boost/scoped_ptr.hpp>
+#include <boost/shared_ptr.hpp>
```
Comment (aerorahul) [inline src/soca/Transforms/HorizConv/HorizConv.h:14]: unused.

File: src/soca/Transforms/HorizConv/HorizConv.h [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+ * This software is licensed under the terms of the Apache Licence Version 2.0
+ * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+#ifndef SOCA_TRANSFORMS_HORIZCONV_HORIZCONV_H_
+#define SOCA_TRANSFORMS_HORIZCONV_HORIZCONV_H_
+
+#include <ostream>
+#include <string>
+#include <boost/scoped_ptr.hpp>
```
Comment (aerorahul) [inline src/soca/Transforms/HorizConv/HorizConv.h:13]: unused.

File: src/soca/Transforms/instantiateBalanceOpFactory.h [hunk 11]
```diff
@@ -34,6 +35,9 @@ void instantiateBalanceOpFactory() {
   static oops::LinearVariableChangeMaker<soca::Traits,
               oops::LinearVariableChange<soca::Traits, soca::Balance> >
               makerBalanceOpBalanceSOCA_("BalanceSOCA");
+  static oops::LinearVariableChangeMaker<soca::Traits,
+              oops::LinearVariableChange<soca::Traits, soca::HorizConv> >
+    makerBalanceOpHorizConvSOCA_("HorizConvSOCA");
```
Comment (aerorahul) [inline src/soca/Transforms/instantiateBalanceOpFactory.h:40]: check alignment

File: test/testinput/dirac_soca_cov_2dfilt.yml [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    lsqrt: 1
+    mpicom: 2
+    date: *date
+    variables: [tocn]
+    inputVariables:
+      variables: *soca_vars #[tocn, socn]
+    outputVariables:
+      variables: *soca_vars #[tocn, socn]
+
+dirac:
```
Comment (aerorahul) [inline test/testinput/dirac_soca_cov_2dfilt.yml:45]: someday, I'ld like to know what these do.

**Review by aerorahul — CHANGES_REQUESTED** [review]
minor comments. otherwise looks good.

**Comment by guillaumevernieres** [comment]
Sorry @aerorahul , closing the pr, since my assumption that bump will do filtering was wrong ... oops.
I'll work on an alternative.

**Pull request closed without merging**